File: ios/EmojiSheetContentView.swift
import ExpoModulesCore
import UIKit

// MARK: - Data Models

struct EmojiItem: Sendable {
    let emoji: String
    let name: String
    let version: String
    let toneEnabled: Bool
    let keywords: [String]
    let id: String
}

struct EmojiSection: Sendable {
    let title: String
    var data: [EmojiItem]
}

// MARK: - Content View (ExpoView wrapper)

class EmojiSheetContentView: ExpoView, EmojiSheetUIViewDelegate {
    let onEmojiSelected = EventDispatcher()
    let onDismiss = EventDispatcher()
    let onOpen = EventDispatcher()
    private let pickerView = EmojiSheetUIView()

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        pickerView.translatesAutoresizingMaskIntoConstraints = false
        pickerView.delegate = self
        addSubview(pickerView)
        NSLayoutConstraint.activate([
            pickerView.topAnchor.constraint(equalTo: topAnchor),
            pickerView.leadingAnchor.constraint(equalTo: leadingAnchor),
            pickerView.trailingAnchor.constraint(equalTo: trailingAnchor),
            pickerView.bottomAnchor.constraint(equalTo: bottomAnchor),
        ])
        pickerView.loadDataAsync()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func didMoveToWindow() {
        super.didMoveToWindow()
        if window != nil {
            onOpen([:])
        }
    }

    func updateTheme(_ theme: String) {
        pickerView.updateTheme(theme)
    }

    func updateCategoryBarPosition(_ position: String) {
        pickerView.categoryBarPosition = position
    }

    func updateLayoutDirection(_ direction: String) {
        pickerView.layoutDirection = direction
    }

    func updateColumns(_ columns: Int) {
        pickerView.columns = columns
    }

    func updateEmojiSize(_ size: CGFloat) {
        pickerView.emojiSize = size
    }

    func updateRecentLimit(_ limit: Int) {
        pickerView.recentLimit = limit
    }

    func updateShowSearch(_ show: Bool) {
        pickerView.showSearch = show
    }

    func updateShowRecents(_ show: Bool) {
        pickerView.showRecents = show
    }

    func updateEnableSkinTones(_ enable: Bool) {
        pickerView.enableSkinTones = enable
    }

    func updateEnableHaptics(_ enable: Bool) {
        pickerView.enableHaptics = enable
    }

    func updateEnableAnimations(_ enable: Bool) {
        pickerView.enableAnimations = enable
    }

    func updateSearchPlaceholder(_ text: String) {
        pickerView.searchPlaceholder = text
    }

    func updateNoResultsText(_ text: String) {
        pickerView.noResultsText = text
    }

    func updateCategoryNames(_ names: [String: String]) {
        pickerView.categoryNames = names
    }

    func updateExcludeEmojis(_ ids: [String]) {
        pickerView.excludeEmojis = Set(ids)
    }

    // MARK: - EmojiSheetUIViewDelegate

    func emojiSheetView(_ view: EmojiSheetUIView, didSelectEmoji data: [String: Any]) {
        onEmojiSelected(data)
    }
}

File: android/src/main/java/expo/community/modules/emojisheet/EmojiSheetContentView.kt
package expo.community.modules.emojisheet

import android.annotation.SuppressLint
import android.content.Context
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

@SuppressLint("ViewConstructor")
class EmojiSheetContentView(
    context: Context,
    appContext: AppContext
) : ExpoView(context, appContext) {

    private val onEmojiSelected by EventDispatcher()
    private val pickerView = EmojiSheetUIView(context)

    init {
        pickerView.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
        pickerView.onEmojiSelected = { data ->
            onEmojiSelected(data)
        }
        addView(pickerView)
        pickerView.loadDataAsync()
    }

    fun applyConfiguration() {
        pickerView.applyConfiguration()
    }

    fun updateTheme(theme: String) {
        pickerView.updateTheme(theme)
    }

    fun updateCategoryBarPosition(position: String) {
        pickerView.categoryBarPosition = position
        pickerView.applyConfiguration()
    }

    fun updateColumns(columns: Int) {
        pickerView.columns = columns
        pickerView.applyConfiguration()
    }

    fun updateEmojiSize(size: Float) {
        pickerView.emojiSize = size
        pickerView.applyConfiguration()
    }

    fun updateRecentLimit(limit: Int) {
        pickerView.recentLimit = limit
    }

    fun updateShowSearch(show: Boolean) {
        pickerView.showSearch = show
    }

    fun updateShowRecents(show: Boolean) {
        pickerView.showRecents = show
    }

    fun updateEnableSkinTones(enable: Boolean) {
        pickerView.enableSkinTones = enable
    }

    fun updateEnableHaptics(enable: Boolean) {
        pickerView.enableHaptics = enable
    }

    fun updateSearchPlaceholder(text: String) {
        pickerView.searchPlaceholder = text
    }

    fun updateNoResultsText(text: String) {
        pickerView.noResultsText = text
    }

    fun updateCategoryNames(names: Map<String, String>) {
        pickerView.categoryNames = names
    }

    fun updateExcludeEmojis(ids: List<String>) {
        pickerView.excludeEmojis = ids.toSet()
    }
}
